pandas/io/formats/latex.py

```diff
@@ -256,88 +254,394 @@ def _format_multirow(
+        """Create iterator over header or body of the table.
 
         Parameters
         ----------
-        buf : string or file handle
-            File path or object. If not specified, the result is returned as
-            a string.
-        column_format : str
-            The columns format as specified in `LaTeX table format
-            <https://en.wikibooks.org/wiki/LaTeX/Tables>`__ e.g 'rcl'
-            for 3 columns
+        over : {'body', 'header'}
+            Over what to iterate.
+
+        Returns
+        -------
+        RowStringConverter
+            Iterator over body or header.
         """
-        if self._table_float:
-            # then write output in a nested table/tabular or longtable environment
-            if self.caption is None:
-                caption_ = ""
-            else:
-                caption_ = f"\n\\caption{{{self.caption}}}"
+        kwargs = dict(
+            formatter=self.fmt,
+            multicolumn=self.multicolumn,
+            multicolumn_format=self.multicolumn_format,
+            multirow=self.multirow,
+        )
+        if over == "header":
+            return RowHeaderIterator(**kwargs)
+        elif over == "body":
+            return RowBodyIterator(**kwargs)
+        else:
+            msg = f"'over' must be either 'header' or 'body', but {over} was provided"
+            raise ValueError(msg)
+
+
+class LongTableBuilder(TableBuilder):
+    """Concrete table builder for longtable."""
+
+    @property
+    def env_begin(self):
+        first_row = (
+            f"\\begin{{longtable}}{self._position_macro}{{{self.column_format}}}"
+        )
+        elements = [first_row, f"{self._caption_and_label()}"]
+        return "\n".join([item for item in elements if item])
+
+    def _caption_and_label(self):
+        if self.caption or self.label:
+            double_backslash = "\\\\"
+            elements = [f"{self._caption_macro}", f"{self._label_macro}"]
+            caption_and_label = "\n".join([item for item in elements if item])
+            caption_and_label += double_backslash
+            return caption_and_label
+        else:
+            return ""
+
+    @property
+    def middle_separator(self):
+        iterator = self._create_row_iterator(over="header")
+        elements = [
+            "\\midrule",
+            "\\endhead",
+            "\\midrule",
+            f"\\multicolumn{{{len(iterator.strcols)}}}{{r}}"
+            "{{Continued on next page}} \\\\",
+            "\\midrule",
+            "\\endfoot\n",
+            "\\bottomrule",
+            "\\endlastfoot",
+        ]
+        if self._is_separator_required():
+            return "\n".join(elements)
+        return ""
+
+    @property
+    def bottom_separator(self):
+        return ""
+
+    @property
+    def env_end(self):
+        return "\\end{longtable}"
+
+
+class RegularTableBuilder(TableBuilder):
+    """Concrete table builder for regular table."""
+
+    @property
+    def env_begin(self):
```

Comment:
try to type *everything*

Reply:
Did that.
But faced mypy errors, that I do not know how to overcome.

``
pandas/io/formats/latex.py:44: error: Argument 1 to "list" has incompatible type "Iterator[Tuple[Any, ...]]"; expected "Iterable[List[str]]"  [arg-type]
pandas/io/formats/latex.py:544: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]
``
mypy does not know, for some reason, that list can get an iterator rather than iterable.

Another thing is related to ``column_format`` setter, that I do not know how to deal with.

Can you help?